bot/event_looper.py

```diff
@@ -0,0 +1,271 @@
+import datetime
+import logging
+import re
+import sqlite3
+import threading
+from typing import Optional
+
+import schedule
+import time
+from ts3.response import TS3Event
+
+from bot.TS3Auth import AuthRequest
+from bot.db import ThreadSafeDBConnection
+from bot.ts import Channel, TS3Facade, User
+from .audit_service import AuditService
+from .config import Config
+from .connection_pool import ConnectionPool
+from .user_service import UserService
+
+LOG = logging.getLogger(__name__)
+
+
+class EventLooper:
+    def __init__(self, database_connection: ThreadSafeDBConnection,
+                 ts_connection_pool: ConnectionPool[TS3Facade],
+                 config: Config,
+                 user_service: UserService,
+                 audit_service: AuditService):
+        self._database_connection = database_connection
+        self._ts_connection_pool = ts_connection_pool
+        self._config = config
+        self._user_service = user_service
+        self._audit_service = audit_service
+
+        self._lock = threading.RLock()
+
+        self._ts_facade: Optional[TS3Facade] = None
+        self._own_client_id = None
+
+    def start(self):
+        with self._lock:  # prevent concurrency
+            with self._ts_connection_pool.item() as ts_facade:
+                self._ts_facade = ts_facade
+
+                self._set_up_connection()
+
+                LOG.info("BOT now online,sending broadcast.")
+                self._broadcast_message()  # Send initial message into channel
+
+                # Forces script to loop forever while we wait for events to come in, unless connection timed out or exception occurs.
+                # Then it should loop a new bot into creation.
+                LOG.info("BOT now idle, waiting for requests.")
+
+                self._loop_for_events()
+
+    def _loop_for_events(self):
+        while self._ts_facade.is_connected():
```

Comment:
I'm not 100% happy with the calss, as some class variables should probably be better handled as parameters in this call stack OR handled with instances of smaller scoped "Worker" Classes.

Reply:
Understandably so. Since I have switched sides into the FP camp through the more recent years, I would advocate for passing all required information through as parameters to facilitate referential transparency, but I can see how that can be tedious for deep call stacks and we are in OOP after all. So maybe sparse use of singleton for those three classes is in order?

Reply:
Not sure how this would look like ?
Fow now I can add a NoneCheck in the while condition

Reply:
Yes, that's exactly what I had in mind.